infrastructure/src/main/java/org/corfudb/infrastructure/BatchWriter.java

```diff
@@ -0,0 +1,126 @@
+package org.corfudb.infrastructure;
+
+import com.github.benmanes.caffeine.cache.CacheWriter;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.log.LogAddress;
+import org.corfudb.infrastructure.log.StreamLog;
+import org.corfudb.protocols.wireprotocol.LogData;
+import org.corfudb.runtime.exceptions.OverwriteException;
+
+import javax.annotation.Nonnull;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * Created by m on 11/30/16.
+ */
+@Slf4j
+public class BatchWriter <K, V> implements CacheWriter<K, V> {
+
+    private StreamLog streamLog;
+    private BlockingQueue<WriteOperation> writeQueue;
+    final ExecutorService writerService = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+            .setDaemon(false)
+            .setNameFormat("LogUnit-Write-Processor-%d")
+            .build());
+
+    public BatchWriter(StreamLog streamLog) {
+        this.streamLog = streamLog;
+        writeQueue = new LinkedBlockingQueue<>();
+        writerService.submit(this::batchWriteProcessor);
+    }
+
+    @Override
+    public void write(@Nonnull K key, @Nonnull V value) {
+        try {
+            submitWrite((LogAddress) key, (LogData) value).get();
+        } catch (Exception e) {
+            if(e.getCause() instanceof OverwriteException) {
+                throw new OverwriteException();
+            } else {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    @Override
+    public void delete(K key, V value, RemovalCause removalCause) {}
+
+    private CompletableFuture<Void> submitWrite(LogAddress address, LogData logData) {
+        CompletableFuture<Void> cf = new CompletableFuture();
+        writeQueue.add(new WriteOperation(address, logData,cf));
+        return cf;
+    }
+
+    private void batchWriteProcessor() {
+        try {
+            WriteOperation lastOp = null;
+            int batchSize = 100;
+            int processed = 0;
+
+            List<CompletableFuture> ack = new LinkedList();
+            List<CompletableFuture> err = new LinkedList();
+
+            while (true) {
+                WriteOperation currOp;
+
+                if (lastOp == null) {
+                    currOp = writeQueue.take();
```

Comment:
I am puzzled. This is a blocking call. In case of a low load, this means we will not be able to get a future-completion until a batch is filled up?